packages/runtime-common/virtual-network.ts

```diff
@@ -86,3 +205,63 @@ function isUrlLike(moduleIdentifier: string): boolean {
     moduleIdentifier.startsWith('https://')
   );
 }
+
+async function getContentOfReadableStream(
+  requestBody: ReadableStream<Uint8Array> | null,
+): Promise<Uint8Array | null> {
+  if (requestBody) {
+    let isPending = true;
+    let arrayLegnth = 0;
+    let unit8Arrays = [];
+    let reader = requestBody.getReader();
+    do {
+      let readableResults = await reader.read();
+
+      if (readableResults.value) {
+        arrayLegnth += readableResults.value.length;
+        unit8Arrays.push(readableResults.value);
+      }
+
+      isPending = !readableResults.done;
+    } while (isPending);
+    let mergedArray = new Uint8Array(arrayLegnth);
+    unit8Arrays.forEach((array) => mergedArray.set(array));
+    return mergedArray;
+  }
+  return null;
+}
+
+async function buildRequest(url: string, originalRequest: Request) {
+  if (url === originalRequest.url) {
+    return originalRequest;
+  }
+
+  // To reach the goal of creating a new Request but with a different url it is
+  // usually enough to create a new Request object with the new url and the same
+  // properties as the original request, but there are issues when the body is
+  // a ReadableStream - browser reports the following error:
+  // "TypeError: Failed to construct 'Request': The `duplex` member must be
+  // specified for a request with a streaming body." Even adding the `duplex`
+  // property will not fix the issue - the browser request being made to
+  // our local server then expects HTTP/2 connection which is currently not
+  // supported in our local server. To avoid all these issues, we resort to
+  // reading the body of the original request and creating a new Request with
+  // the new url and the body as a Uint8Array.
+
+  let body = null;
+  if (originalRequest.body) {
+    body = await getContentOfReadableStream(originalRequest.clone().body);
```

Comment:
I've been trying to see if I can get rid of body stream reading to create a new request with a changed url but to no avail...

1. `Object.defineProperty(request, 'url', { value: mappedUrl })` will not work because the underlying url value really is immutable. Here is a demonstration - the request will still try to reach the original url even when trying to override it with `defineProperty`: 

<img width="579" alt="image" src="https://github.com/cardstack/boxel/assets/273660/90d5cd86-852b-4c28-829b-bd214e8edc8d">

2. I saw some hints online that [tee](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/tee) could be used to copy the body, like so:

```
let copiedBody = null;
  if (originalRequest.body) {
    copiedBody = originalRequest.body.tee()[0];
  }

  return new Request(url, {
    method: originalRequest.method,
    headers: originalRequest.headers,
    body: copiedBody,
    referrer: originalRequest.referrer,
    referrerPolicy: originalRequest.referrerPolicy,
    mode: originalRequest.mode,
    credentials: originalRequest.credentials,
    cache: originalRequest.cache,
    redirect: originalRequest.redirect,
    integrity: originalRequest.integrity,
  });
  ```
  
  
  But again, the error persists:
  
<img width="1170" alt="image" src="https://github.com/cardstack/boxel/assets/273660/88c3efb1-021f-49da-9c7f-16ae69663964">

3. `clone()`-ing the request and taking its body won't work either (failing with the same error): 

```
return new Request(url, {
    method: originalRequest.method,
    headers: originalRequest.headers,
    body: originalRequest.clone().body,
    referrer: originalRequest.referrer,
    referrerPolicy: originalRequest.referrerPolicy,
    mode: originalRequest.mode,
    credentials: originalRequest.credentials,
    cache: originalRequest.cache,
    redirect: originalRequest.redirect,
    integrity: originalRequest.integrity,
  });
```

It is interesting because Safari gives a different error message:

<img width="415" alt="image" src="https://github.com/cardstack/boxel/assets/273660/e1ce8d02-e6f2-4a26-bff8-6ed938541391">

Reply:
Another thought would be to Proxy the request: 
```ts 
let proxiedRequest = new Proxy(originalRequest, {
  get(target, property, received) {
    if (property === 'url') {
      return mappedURL;
    }
    return Reflect.get(target, property, received);
  }
})